all: Handling boarding / alighting duration in navitia

eturck opened the pull request:
Hello! A few months later than it should have been, but here is the PR about boarding times, feel free to tag this PR as "do not merge" for the moment, I don't think I can do it myself. Quick summary :
* It's following the issue #1662,
* The goal is to be able to use boarding and alighting duration in navitia. It may represents boat or plane boarding or weird on demand transport modelisation :). The NTFS spec added the two fields since #1654,
* There is one change in navitia-proto that needs to be merged (adding a section_type in response) : see PR CanalTP/navitia-proto#84,
* When calling journeys it's adding boarding / alighting sections working the same way than boarding / landing sections existing right now, so there is no changes on the jormungandr side (except more integration tests),
* There are a few places where we were considering that the smallest time of a vj was the arrival_time on the first stop_time, it's not the case anymore since any boarding_time could be smaller,
* I didn't change the output of route_schedules / stop_schedules / departure / arrival at all, it's still showing departure / arrival times (just made some code changes since raptor data have boarding / alighting times now). I don't know if you want to add some information on boarding / alighting in there? 

The only big problem I encountered was the non chronological boarding / alighting times messing with the shifting of datetimes. If we have this in the NTFS :

| sequence | departure_time | arrival_time | boarding_duration | alighting_duration |
| --- | --- | --- | --- | --- |
| 0 | 08:00:00 | 08:00:00 | 0 | 0 |
| 1 | 08:05:00 | 08:04:00 | 0 | 15 |
| 2 | 08:10:00 | 08:08:00 | 15 | 5 |

When computing boarding and alighting_time we have : 

| order | boarding_time | departure_time | arrival_time | alighting_time |
| --- | --- | --- | --- | --- |
| 0 | 08:00:00 | 08:00:00 | 08:00:00 | 08:00:00 |
| 1 | 08:05:00 | 08:05:00 | 08:04:00 | 08:19:00 |
| 2 | 07:55:00 | 08:10:00 | 08:08:00 | 08:13:00 |

It doesn't bother raptor at all since we still can't go back in time, but the shifting done with datetime::shift was not working, because when searching clockwise we expect an arrival later for the next stop_time, otherwise we consider there is a shift in the day and with alighting it's not always the case (as we can see for stop_time 1 and 2 above).
Now I keep track of the current shift in raptor and call section_end and the other function working with the stop_time with non shifted times. I don't know if it's the most elegant way to do it, waiting your feedback on this one :). (most of the changes about this are on the commit bf7b239)

Let me know what you think about all of this!

### Quick fixes I've done that may be separated in other PR ?

I used alembic to autogenerate the version : 
```bash
PYTHONPATH=. alembic -x dbname="postgresql://user:password@localhost/navitia" revision --autogenerate -m "message"
```
But it was creating a lot more things than it was supposed to do. The navitia.py model wasn't up to date. Maybe I didn't do it the right way, or something else didn't :)? I fixed it in a90fae9

----------------

In 81cd4dc I've changed a frequency vj stop_time departure time from 1000 to 87400 because the first one was at 70000. It was breaking some of my changes and I don't think it was valid? If I'm mistaken let me know because it's breaking boarding_time on frequencies.

Commit:
sql: fix navitia model in alembic

There was multiple database upgrades that weren't reflected in the python
navitia model.

Commit:
sql: add boarding and alighting time to stop_time table

Commit:
ed: import boarding and alighting duration from NTFS

We have two new optional fields in NTFS:
  - boarding_duration,
  - alighting_duration.
Representing the necessary time taken by the boarding process or
alighting process. We store the boarding and alighting time into
database. The boarding_time is the departure_time minus the boarding
duration and the alighting is the arrival_time plus the alighting
duration. These values can't be negative.

Commit:
ed: fix shift_stop_times test after boarding / alighting addition

Commit:
ed: add boarding/alighting times in fusio_parser_test

Commit:
type: add boarding and alighting time in navitia::type

Commit:
all: use boarding and alighting time in place of departure arrival

Replace departure_time / arrival_time in the routing by boarding and
alighting_time.
The main issue is that we can't use datetime::shift anymore because it
was supposing that the departure_time were ordered chronologicaly when
moving to the next stop_time, otherwise, it was a day shift. It's not
true anymore since the boarding duration at a stop could vary. Therefore
we maintain the shift time directly in raptor and only work with non
shifted times.

Commit:
routing: output boarding / alighting duration section

When we have a boarding or alighting time different than the
departure_time / arrival_time at the start / end of a section (not
including stay in vj) output a boarding or alighting section. Use the
same type of section than the boarding / landing section existing right
now.

Commit:
ed: add boarding and alighting duration to builder

Commit:
routing: add unit tests for boarding and alighting times

Commit:
routing: fix test data on get_stop_times

Commit:
kraken: handle boarding / alighting time on disruptions

Commit:
kraken: unit test for boarding and alighting times on disruptions

Commit:
time_tables: changes to still use departure and arrival times

Since the different services are using raptor datas we need to add or
substract boarding / alighting duration to get the original datas.

Commit:
time_tables: unit tests with boarding and alighting times

Commit:
jormungandr: integration tests for boarding alighting sections

pbougue commented:
> I didn't change the output of route_schedules / stop_schedules / departure / arrival at all, it's still showing departure / arrival times (just made some code changes since raptor data have boarding / alighting times now). I don't know if you want to add some information on boarding / alighting in there?

We discussed it, and to us it's OK that way for now.
At some point we will have to output it, but you might be the first one to need it.
Probably simply adding the boarding time to `/departures` `/route_schedules` and `/stop_schedules`, and alighting time to `/arrivals`.
So this can be done in another PR (and we all agree that this one doesn't need to inflate more :wink: )

TeXitoi commented on source/kraken/apply_disruption.cpp:
```diff
@@ -175,6 +176,14 @@ struct add_impacts_visitor : public apply_impacts_visitor {
+                    st.boarding_time = st.departure_time - base_st->get_boarding_duration();
+                    st.alighting_time = st.arrival_time + base_st->get_alighting_duration();
+                }
+            }
```
It can't work: `auto st` is doing a copy of the stop_time, and then you try to modify it.

Maybe a test is needed to check this code.

eturck replied:
Yes, definitely needs a test. I completely missed it, I thought I had it but I only tested disruption on line_sections.

pbougue commented on source/ed/build_helper.cpp:
```diff
@@ -730,11 +733,18 @@ void builder::finish() {
              }
 
-             const auto start = freq_vj->stop_time_list.front().arrival_time;
+             const auto first_st_it = freq_vj->earliest_stop_time();
```
Maybe add some comment on why that `freq_vj->start_time` shift is necessary (only to have a first stop at the good date?).
Also why not shifting end_time based on alighting (or why even shifting it at all)?

This code is duplicated, may deserve it's own method in vj.

eturck replied:
That's a question I forgot to ask!

I don't know if shifting start_time is necessary or not. The thinking behind was if you have a frequency vj starting at 08:00:00, you're going to set a start_time at 08:00:00 in the NTFS, even if the boarding on some st is XX minutes, and the boarding_time is prior to 08:00:00.
But for the routing we need the start_time at the smallest possible boarding_time / arrival_time. So start_time should be at 07:XX:00.
There is two ways : 
* The start_time in the NTFS should take that into account, and I remove this code,
* We need to take care about it.

And that's why I shift end_time the same way. Currently end_time represents the last possible time at the first stop_time. Now end_time represents the last possible boarding_time.

TeXitoi commented on source/routing/get_stop_times.cpp:
```diff
@@ -60,9 +60,9 @@ std::vector<datetime_stop_time> get_stop_times(const routing::StopEvent stop_eve
         if (st.first) {
-            next_requested_dt.push({jpp_idx, st.first, st.second});
+            auto diff = clockwise ? st.first->get_boarding_duration() : -1 * st.first->get_alighting_duration();
+            next_requested_dt.push({jpp_idx, st.first, st.second + diff});
```
That's a bit strange: we will get the departures ordered/extracted by boarding_datetime, but we will print the departure_datetime.

eturck replied:
Needs a test too I guess :). On departure_boards there is a sort after calling get_stop_times but it doesn't seems to be the case everywhere..

eturck replied:
This part of my code was vaguely stupid... I added some unit tests and I'm going to add more. And while stop_schedules and route_schedules are sorting their departures, passages wasn't. Thanks for the heads up on that one :).

TeXitoi commented on source/routing/get_stop_times.cpp:
```diff
@@ -82,7 +82,8 @@ std::vector<datetime_stop_time> get_stop_times(const routing::StopEvent stop_eve
         if (st.first) {
-            next_requested_dt.push({best_jpp_dt.jpp, st.first, st.second});
+            auto diff = clockwise ? st.first->get_boarding_duration() : -1 * st.first->get_alighting_duration();
+            next_requested_dt.push({best_jpp_dt.jpp, st.first, st.second + diff});
```
can this code be mutualized with l64?

TeXitoi commented on source/routing/get_stop_times.cpp:
```diff
@@ -181,12 +182,12 @@ get_all_calendar_stop_times(const routing::JourneyPattern& jp,
+                return t > freq_vj->end_time + st.departure_time;
             };
-            for (auto time = freq_vj->start_time; stop_loop(time); time += freq_vj->headway_secs) {
+            for (auto time = freq_vj->start_time + st.departure_time; stop_loop(time); time += freq_vj->headway_secs) {
```
Is this fix related to boarding/alighting?

eturck replied:
I saw something wrong while testing frequencies with boarding / alighting (since the start_time could be the first boarding_time).
But I think if we currently have the arrival_time of the first stop_time smaller than its departure_time it would break too, because we assume the start_time is the departure_time but it's based on the arrival_time.

pbougue commented on source/ed/data.cpp:
```diff
@@ -205,9 +204,11 @@ void Data::shift_stop_times() {
         //
         // For frequency vj, the start time must be in [0; 24:00[.
-        const int start_time = vj->is_frequency() ? vj->start_time : first_st->arrival_time;
+        const int start_time = vj->is_frequency() ? vj->start_time : std::min(first_st->boarding_time, first_st->arrival_time);
```
minor: line length

eturck replied:
What is the limit? I thought it was 130.

pbougue replied:
It's 113 (you can check it on github, tab files changed, split view)

eturck replied:
It's screen resolution dependent no? The line seems ok on a full HD monitor (not so much on a WXGA one though =p). But that's not really important, I'm going to set the limit at 113 on my IDE and change all this :).

TeXitoi commented on source/routing/raptor.cpp:
```diff
@@ -63,18 +63,20 @@ bool RAPTOR::apply_vj_extension(const Visitor& v,
         const auto& st_begin = stop_time_list.front();
-        workingDt = st_begin.section_end(workingDt, v.clockwise());
+        shift = st_begin.current_shift_time(workingDt, v.clockwise());
+        workingDt = st_begin.section_end(workingDt - shift, v.clockwise()) + shift;
```
I don't like that, but I have to think about it to find a better alternative.

TeXitoi commented on source/routing/raptor.cpp:
```diff
@@ -692,10 +692,9 @@ void RAPTOR::raptor_loop(Visitor visitor,
                     const DateTime previous_dt = prec_labels.dt_transfer(jpp.sp_idx);
                     if (prec_labels.transfer_is_initialized(jpp.sp_idx) && valid_stop_points[jpp.sp_idx.val] &&
-                        (!is_onboard || visitor.better_or_equal(previous_dt, workingDt, *it_st))) {
+                        (!is_onboard || visitor.better_or_equal(previous_dt, workingDt, *it_st, shift))) {
```
Why do you need shift here?

eturck replied:
Because better_or_equal calls section_end which was one of the function shifting. But I guess I could do
```
visitor.better_or_equal(previous_dt - shift, workingDt - shift, *it_st)))
```
With the same result (I we keep the shifting ;)). I don't know why I didn't thought of this.

TeXitoi commented on source/routing/raptor_solution_reader.cpp:
```diff
@@ -727,8 +733,22 @@ Path make_path(const Journey& journey, const type::Data& data) {
+            auto& vj_section = sections.at(i);
+            const auto& boarding_st = vj_section.stop_times_and_dt.front().st;
+            const auto& alighting_st = vj_section.stop_times_and_dt.back().st;
+            if (i == 0 && boarding_st.boarding_time != boarding_st.departure_time) {
```
Can't we manage that outside of the loop?

eturck replied:
We definitely can... I'm dumb.

TeXitoi commented on source/routing/raptor_solution_reader.cpp:
```diff
@@ -761,6 +782,18 @@ Path make_path(const Journey& journey, const type::Data& data) {
             item.arrival = posix(vj_section.stop_times_and_dt.back().arrival);
             last_vj_section = &vj_section;
+
+            if (i == sections.size() - 1 && alighting_st.alighting_time != alighting_st.arrival_time) {
```
Can't we manage that outside of the loop?

TeXitoi commented on source/routing/raptor_visitors.h:
```diff
@@ -9,8 +9,8 @@ struct raptor_visitor {
-    inline bool better_or_equal(const DateTime& a, const DateTime& current_dt, const type::StopTime& st) const {
-        return a <= st.section_end(current_dt, clockwise());
+    inline bool better_or_equal(const DateTime& a, const DateTime& current_dt, const type::StopTime& st, const DateTime& shift) const {
+        return a <= (st.section_end(current_dt - shift, clockwise()) + shift);
```
I really don't like these shifts...

TeXitoi commented on source/type/type.cpp:
```diff
@@ -554,6 +569,12 @@ const VehicleJourney* VehicleJourney::get_corresponding_base() const {
     return nullptr;
 }
 
+std::vector<StopTime>::const_iterator VehicleJourney::earliest_stop_time() const {
```
Returning an iterator here is not so userfriendly: the end iterator to check is not trivial. Maybe returning a pointer or even a reference if that can't fail is a better idea.

eturck replied:
Can we have a vj with no stop_time? The clean method of data.cpp delete vj with empty stop_time_list, and realtime.cpp seems to not handle realtime message with empty aux_info.stop_times, so I guess not?

TeXitoi reviewed (changes requested):
:+1: Great work!

Everything seems good except this shift thing. Also, I've only lightly read the tests.

I'll think about this shift thing, maybe I have an idea.

pbougue commented on source/routing/tests/routing_api_test_data.h:
```diff
@@ -428,6 +428,10 @@ struct routing_api_data {
             b.lines["C"]->code = "1C";
 
+            b.vj("C")
+                ("stop_point:stopA", "23:00:00"_t, "23:02:00"_t, std::numeric_limits<uint16_t>::max(), false, true, 0, 1800)
```
minor: line length

pbougue commented on source/time_tables/passages.cpp:
```diff
@@ -168,24 +168,28 @@ void passages(PbCreator& pb_creator,
-            passage->mutable_stop_date_time()->set_base_arrival_date_time(navitia::to_posix_timestamp(base_passage_dt));
+            auto base_departure_dt = get_base_dt(dt_stop_time.second, base_st, base_ptime, true);
+            auto base_arrival_dt = get_base_dt(dt_stop_time.second, base_st, base_ptime, false);
+            passage->mutable_stop_date_time()->set_base_departure_date_time(navitia::to_posix_timestamp(base_departure_dt));
```
weird indent

pbougue commented on source/type/type.h:
```diff
@@ -708,6 +710,10 @@ struct StopTime {
     uint32_t departure_time = 0; ///< seconds since midnight
 
+    /// Representing times used in Raptor, taking into account boarding and alighting process duration
+    uint32_t boarding_time = 0; ///< seconds since midnight
```
Adding two int32 to StopTime is a risk for RAM use, we have to measure its impact (and probably Kisio should do that on it's biggest instances)

- [x] Monitor RAM use increase (on fr-idf?)

eturck replied:
Let me know if that's a problem and if I can do something about this :).

pbougue commented on source/routing/tests/get_stop_times_test.cpp:
```diff
@@ -263,7 +263,7 @@ BOOST_AUTO_TEST_CASE(test_looping_frequency_for_calendar) {
     b.frequency_vj("A", vj1_departure, 2001, headway_sec, "default_network", "1010", "", true, "vj1")
         (spa1, vj1_departure, vj1_departure)
-        ("useless_stop", 1000, 1000);
+        ("useless_stop", 87400, 87400);
```
It sounds OK that we break on that... What's weird is it wasn't the case before.

pbougue reviewed (commented):
Thaaaaaat PR :weight_lifting_man: 

Hard to read it all :sweat: 
And didn't read much of the tests so I have some ideas of tests that might break (and that you might already have):
* from experience I'd say we might miss reverse journey computing tests (request an arrival datetime).
* also, one interesting test can be on an NTFS "anti-passing" midnight because of boarding (like first stop_time at 00:05:00 - no boarding, second at 00:10:00 - boarding 30min).

As for the RAM increase, we have to coordinate and test it, you can do some checks if you want, but we'll probably have to test it at Kisio also.

eturck commented:
Thank you for the feedback :).

@pbougue I added only one test to reverse_raptor_test, maybe I should add more. When testing with our zonal ODT it seems to work pretty well, but we don't have some weird pass midnight or things like that. The "anti-passing" midnight one is a good idea, I don't think I have it and it's probably going to break everything ;).

@TeXitoi about the shifting, I tried a bunch of stuff but nothing was really working. I'm sure there is a better solution though. Let me know if you want to know what was breaking when testing things.

Commit:
kraken: add (failing) test on realtime with boarding

TeXitoi commented:
I may have an idea to simplify all these shift things. I can explain that here, and you give me your feedbacks.

The thing is that the times in the stop time are always relative to some datetime. For example, for a normal vj, that's midnight the starting date of the vj. For frequency vj, that something like the departure of the vj.

Thus, the idea would be to store this base datetime when taking a vj. It can be easily calculated as the difference between the datetime given by next_stop_time, and the corresponding field of the stop_time. Then, for freq and normal vj, when we have the next stop time, we just have to do the addition between the field and our base datetime. It would be the same for all the vj, same code for freq and normal, and not clockwise related except for the field used. Instead of giving the previous datetime, or the previous datetime plus the shift, we just have to give this base datetime.

What do you think?

Commit:
kraken: fix boarding time for disruption on meta_vj

xlqian commented on source/ed/connectors/gtfs_parser.cpp:
```diff
@@ -1074,23 +1078,32 @@ void FrequenciesGtfsHandler::handle_line(Data& data, const csv_row& row, bool) {
+        vj->end_time = to_utc(row[end_time_c], utc_offset);
+        vj->headway_secs = boost::lexical_cast<int>(row[headway_secs_c]);
+        const auto it_st = vj->earliest_stop_time();
+        auto start = std::min((*it_st)->boarding_time, (*it_st)->arrival_time);
```
since you used `&` at https://github.com/CanalTP/navitia/pull/1980/files#diff-df43d00a168ef46f7768a38fe8f1de18R478, you can do `const auto&`, too :)

xlqian commented on source/routing/raptor_solution_reader.cpp:
```diff
@@ -727,8 +733,22 @@ Path make_path(const Journey& journey, const type::Data& data) {
+                                        posix(section.get_in_dt + boarding_st.get_boarding_duration()));
+                auto& boarding_section = path.items.back();
+                boarding_section.stop_points.push_back(boarding_st.stop_point);
+                boarding_section.stop_points.push_back(boarding_st.stop_point);
```
@TeXitoi explained to me why L747 is the same as L748... just for the clarity, can you add a comment plz? :)

xlqian commented on source/routing/tests/frequency_raptor_test.cpp:
```diff
@@ -427,3 +427,217 @@ BOOST_AUTO_TEST_CASE(freq_vj_end_time_is_smaller_than_start_time) {
+    result.at(0).print();
+    BOOST_REQUIRE_EQUAL(result.at(0).items.size(), 3);
+    BOOST_CHECK_EQUAL(result.at(0).items[0].type, ItemType::boarding);
+    BOOST_CHECK_EQUAL(result.at(0).items[0].departure, boost::posix_time::time_from_string("2017-Jan-01 08:25:00"));
```
(i don't know if it's possible, if you haven't tried yet) can you do `"20170101T082500"_dt`?

eturck replied:
It works, thanks :).

xlqian commented on source/type/type.cpp:
```diff
@@ -423,7 +429,10 @@ VJ* MetaVehicleJourney::impl_create_vj(const std::string& uri,
     vj_ptr->shift = 0;
     if (!sts.empty()) {
-        vj_ptr->shift = sts.front().arrival_time / (ndtu::SECONDS_PER_DAY);
+        const auto first_st_it = std::min_element(sts.begin(), sts.end(), [](const StopTime& st1, const StopTime& st2) {
```
don't you have a `earliest_stop_time()` already? you can make the other one static

eturck replied:
earliest_stop_time is on a vj, not a vector of stop_times. I should probably move the min_element in a separate function taking a vector and just call this one in types::VehicleJourney::earliest_stop_time, type::VehicleJourney::earliest_stop_time and in impl_create_vj. Should I add it in type_utils.h maybe?

xlqian commented:
GG!

eturck commented:
@TeXitoi I think I understand, I'm not really sure :sweat_smile:

The idea would be in raptor.cpp to store, instead of shift : 
```
base_dt = dt - (clockwise ? stop_time->boarding_time : stop_time->alighting_time)
```
And passing this base_dt to section_end instead of passing the current_dt. Something like : 
```
// In raptor.cpp
next_st.section_end(base_dt, clockwise)

// In type.h
DateTime section_end(DateTime base_dt, bool clockwise) const {
    return base_dt + (clockwise ? boarding_time : alighting_time);
}
```
?

TeXitoi commented:
That's the idea yes. Advantages that I see:
  - simpler API (less dependent of clockwise, less operation, easier to understand)
  - same code for freq and normal vj

But I may miss some drawbacks.

Commit:
routing: line length

Commit:
routing: create alighting / boarding outside of the loop

Commit:
time_tables: add failing test with boarding time

Commit:
routing: fix get_stop_times with boarding

We were using clockwise instead of stop_event and storing the
departure_time inside of next_request_dt instead of result.

Commit:
all: minor github remarks

eturck commented:
First batch of changes:
- [x] added a test for SIGNIFICANT_DELAY with boarding which was failing,
- [x] fixed it :smile:,
- [x] moved the code creating boarding and alighting section outside the vj_section loop,
- [x] added a failing test for route_schedules and departure_board when a boarding_time from a vj was before the boarding_time of another but their departure_times were the other way around,
- [x] fixed it too. The changes I originally made to get_stop_time were really bad,
- [x] changes to line length and a few other remarks.

What's still missing:
- [x] Implementing the suggestion of @TeXitoi in place of the shift,
- [x] Changing earliest_stop_time (duplicate code, avoid returning an iterator),
- [x] Adding a bit more tests to reverse_raptor_test,
- [x] Fixing passages. I already fixed it in my working tree but I would like to add some tests. I don't see an existing unit test for passages. Did I miss it? If not can I add a new one?
- [x] Determining if there is a RAM issue. (it's ok)
- [x] Deciding on the frequency start_time / end_time shifting or not (see https://github.com/CanalTP/navitia/pull/1980#discussion_r102241869),

TeXitoi commented:
> Fixing passages. I already fixed it in my working tree but I would like to add some tests. I don't see an existing unit test for passages. Did I miss it? If not can I add a new one?

I can't find one neither. Of course you can create a new test.

Commit:
all: replace earliest_stop_time with earliest_time

Remove duplicate code in a lot of files. Added a function in type_utils
returning a reference to the earliest stop_time. The vj method just
return the time since it's what we wanted 90% of the time.

Commit:
time_tables: add a new test for passages

This test is failing because one of the vj has a boarding_time before
the other but it's the other way around for the departures.

Commit:
time_tables: reorder departures/arrival after calling get_stop_times

It's necessary because of the boarding / alighting times, since we get
the vj by boarding_time but recompute the departure times.

Commit:
routing: minor, line length

Commit:
ed: handle negative boarding_time

When we substract the boaring duration to the departure_time we might
have a boarding_time smaller than 0. We need to shift every stop_time to
the preceding day, and shift the vp if this happens.

Commit:
routing: add a few more tests for boarding times

Check when the boarding_time is shifting a vj to the preceding day.

pbougue commented on source/routing/tests/reverse_raptor_test.cpp:
```diff
@@ -35,6 +35,12 @@ www.navitia.io
+struct logger_initialized {
+    logger_initialized() { init_logger(); }
+};
+BOOST_GLOBAL_FIXTURE( logger_initialized )
```
please add a `;` at the end of line (required in later versions of boost, even if it warns on double `;` in 1.55)

eturck replied:
Ok :). My GCC was warning me of extra ';', I didn't know it was required for newer version.

pbougue commented on source/time_tables/tests/passages_test.cpp:
```diff
@@ -0,0 +1,103 @@
+struct logger_initialized {
+    logger_initialized()   { init_logger(); }
+};
+BOOST_GLOBAL_FIXTURE( logger_initialized )
```
Please add a `;` at the end here also ;)

pbougue commented on source/type/type_utils.h:
```diff
@@ -79,6 +79,15 @@ static bt::ptime get_base_dt(const nt::StopTime* st_orig, const nt::StopTime* st
     return bt::ptime(validity_pattern_dt_day, boost::posix_time::seconds(hour_of_day_base));
 }
 
+static const nt::StopTime& earliest_stop_time(const std::vector<nt::StopTime>& sts) {
```
as this file is included in type.cpp (which affects pretty much the whole project), it would be nice to put the 4 impl of these functions in a cpp (and remove the static).
avoids unused function warning, and avoids recompiling all when modifying one method

pbougue reviewed (commented):
clang/boost1.61 compiling gave those hints

pbougue commented on fixtures/ed/ntfs/frequencies.txt:
```diff
@@ -0,0 +1,2 @@
+trip_id,start_time,end_time,headway_secs
```
Did you launch `make docker_test` after modifying this fixture ?
It seems to me that it fails there (and probably after): https://github.com/CanalTP/navitia/blob/dev/source/ed/docker_tests/ed_integration_tests.cpp#L122 (having 10 vj instead of 8)

This fixture is linked to multiple tests... Not sure if it's worth refactor tests or just modify tests accordingly (at least it's good to test your new fixture in docker_test).

eturck replied:
Hum, no, I didn't launch docker_test. Sorry about that. I'll look into it.

pbougue replied:
As mentionned on IRC : OK to modify both ntfs fixtures (as tests are common) and maybe a refactor for us after to test only a good ntfs v0.6 dataset (no need for 4 or 5 testing), but that's another PR.

pbougue replied:
:white_check_mark:  Done

pbougue reviewed (changes requested):
On my bench, fr-idf (only ntfs data, no journey call so no raptor cache),
Before: 1828104 (res) 2028964 (virt)
After: 1897496 (res) 2104748 (virt)
Increase: 3.80% (res) 3.74% (virt)
With journey calls at different dates, we stay between 2 and 4%.

@TeXitoi @antoine-de @kinnou02 is it OK for you? it's border to me (like I can't decide)...

kinnou02 commented:
what are the alternatives? 
If we have a solution that doesn't impact the performance and is simple we can look at it.
it's less than 80mb increase. 
I would love to only pay this price if we use boarding time.

Commit:
tests: add missing ;

Commit:
type: move functions of type_utils in cpp file

Commit:
ed: fix docker_test after modifying ntfs fixture

Since ntfs and ntfs_v5 are checked with the same method in docker_test I
changed the ntfs_v5 fixture too.

Commit:
routing: remove the shift and use a base datetime instead

Instead of deducting the shift time from the datetime at the previous
stop_time we know compute the base_dt when boarding on a vj. This
base_dt will be the difference between the current_time and the
boarding_time of the starting stop_time :
- For a discrete vj we will have a base_dt equal to midnight at the
  working date,
- For a frequency vj it will be midnight at the working date + the
  start_time + X * headway

pbougue commented:
OK for the RAM overhead we all agree it's OK for the feature we win :)

pbougue commented on source/type/type_utils.cpp:
```diff
@@ -0,0 +1,103 @@
+#include "type/datetime.h"
+#include "type/message.h"
+#include "type/meta_data.h"
+#include "type/type.pb.h"
```
to be included in type_utils.h (and include `type_utils.h` instead)

pbougue commented on source/type/type_utils.h:
```diff
@@ -28,68 +28,22 @@ IRC #navitia on freenode
 #include <boost/date_time/posix_time/posix_time.hpp>
-#include "type/message.h"
 
 namespace nt = navitia::type;
```
We don't want to propagate namespace aliases (in bare .h it is the case)
So we'll have to explicit namespaces here, and
```
#include "type/type.h"
#include "type/type.pb.h"
```

pbougue reviewed (commented):
:+1: docker tests

Commit:
type: fix includes in type_utils

Commit:
type: remove unused functions and minor fixes

pbougue commented:
Jenkins, test this please

antoine-de commented on source/kraken/tests/realtime_test.cpp:
```diff
@@ -1667,3 +1667,131 @@ BOOST_AUTO_TEST_CASE(ordered_delay_message_test) {
+    BOOST_REQUIRE_EQUAL(vj->stop_time_list.at(1).boarding_time, "08:21"_t);
+    BOOST_REQUIRE_EQUAL(vj->stop_time_list.at(2).stop_point->uri, "stop_point:10");
+    BOOST_REQUIRE_EQUAL(vj->stop_time_list.at(2).departure_time, "08:41"_t);
+    BOOST_REQUIRE_EQUAL(vj->stop_time_list.at(2).alighting_time, "08:40"_t);
```
I don't understand why the alighting isn't 08:41 ? 
It's ok for me that we don't keep the 5mn alighting from the delay, but should the alighting be equal to the arrival then ?

eturck replied:
The arrival_time is 08:40, so the alighting is 08:40. But I meant to check arrival_time on the line above, not departure_time :sweat_smile:. I'll change that.

antoine-de commented on source/routing/tests/frequency_raptor_test.cpp:
```diff
@@ -427,3 +427,217 @@ BOOST_AUTO_TEST_CASE(freq_vj_end_time_is_smaller_than_start_time) {
+BOOST_AUTO_TEST_CASE(freq_vj_with_boarding_alighting) {
+    ed::builder b("20170101");
+    b.frequency_vj("A", "8:00"_t, "18:00"_t, "00:30"_t)
+        ("stop1", "8:02"_t, "8:00"_t, std::numeric_limits<uint16_t>::max(), false, true, 0, 300)
```
aren't the departure/arrival switched ?

eturck replied:
You're right they are. I did the same thing in passages_test too.

antoine-de commented on source/routing/tests/frequency_raptor_test.cpp:
```diff
@@ -427,3 +427,217 @@ BOOST_AUTO_TEST_CASE(freq_vj_end_time_is_smaller_than_start_time) {
+    );
+
+    BOOST_REQUIRE_EQUAL(result.size(), 1);
+    result.at(0).print();
```
debug line no ?

eturck replied:
I left them everywhere, my bad.

antoine-de commented on source/routing/tests/frequency_raptor_test.cpp:
```diff
@@ -427,3 +427,217 @@ BOOST_AUTO_TEST_CASE(freq_vj_end_time_is_smaller_than_start_time) {
+    );
+
+    BOOST_REQUIRE_EQUAL(result.size(), 1);
+    result.at(0).print();
```
debug line no ?

antoine-de commented on source/routing/tests/frequency_raptor_test.cpp:
```diff
@@ -427,3 +427,217 @@ BOOST_AUTO_TEST_CASE(freq_vj_end_time_is_smaller_than_start_time) {
+    BOOST_CHECK_EQUAL(result.at(0).items[6].arrival, "20170101T163500"_dt);
+}
+
+BOOST_AUTO_TEST_CASE(transfer_pass_midnight_freq_vj_to_vj_with_boarding_time) {
```
:laughing: :+1:

antoine-de reviewed (commented):
great tests!

Commit:
tests: remove debug traces, fix some errors
